src/server/server.zig

```diff
@@ -23,12 +25,62 @@ const command = @import("command/_command.zig");
 pub const WorldEditData = struct {
 	selectionPosition1: ?Vec3i = null,
 	selectionPosition2: ?Vec3i = null,
-	clipboard: ?main.blueprint.Blueprint = null,
+	clipboard: ?Blueprint = null,
+	undoHistory: History = .{},
+
+	const History = struct {
+		const DLList = std.DoublyLinkedList(Value);
```

Comment:
Linked lists are rather inefficient in terms of memory usage and performance, since every value has to be allocated.
`main..utils.CircularBufferQueue` should work fine. And you also wouldn't need to explicitly handle the node allocations everywhere.

Reply:
And also, I'm not a fan of the alias here, it's only used twice, and one of the use cases would go away with my proposal even.